File: SchoolLms.Domain/StudentStatuses.cs
namespace SchoolLms.Domain;

// ===========================================================================
//  O'quvchi holati (status tag) — docs/modules/students-parity.md §2.3 (S-5).
// ===========================================================================
//
//  NEGA KERAK BO'LDI
//  -----------------
//  EduSchool o'quvchilar ro'yxatida rangli HOLAT ustuni bor va u to'g'ridan
//  to'g'ri satrdan almashtiriladi ("VIP", "Sinov muddatida", "Ko'chib ketmoqchi").
//  Bizda bunday narsa yo'q: o'quvchi yo faol, yo arxivda. Holat — arxivlashdan
//  OLDINGI kuzatuv vositasi, ya'ni boshqa o'q.
//
//  KATALOG SEED QILINMAYDI — ATAYLAB
//  ---------------------------------
//  §2.3 EduSchool'ning sukut holatlari borligini yozadi, LEKIN ularning
//  nomlari bundle'da yo'q (tarjimalar alohida). Nom to'qib chiqarish —
//  maktabning ish jarayonini bizning taxminimiz bilan almashtirish bo'lardi
//  (`certificate_types` da ham aynan shu sabab bilan seed yo'q). Birinchi
//  holatni maktab Sozlamalarda o'zi kiritadi.
//
//  `IsDefault` — o'chirib bo'lmaydigan tizim qatori uchun. Bugun bunday qator
//  yo'q; mijoz sukut holatlar ro'yxatini bersa, ular shu bayroq bilan keladi
//  va ekran ularni tahrirlashni bloklaydi (§2.3.1).
// ===========================================================================

/// <summary>
/// O'quvchi holati — maktab o'zi boshqaradigan rangli katalog (S-5).
/// </summary>
public class StudentStatus
{
    public Guid Id { get; set; } = Guid.NewGuid();

    /// <summary>Ko'rsatiladigan nom. Unikal.</summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>Nishon rangi <c>#RRGGBB</c> ko'rinishida. null = neytral rang.</summary>
    public string? Color { get; set; }

    /// <summary>Ro'yxatdagi tartib (kichikdan kattaga).</summary>
    public int Position { get; set; }

    /// <summary>
    /// Tizim qatori — tahrirlab va o'chirib bo'lmaydi (EduSchool'dagi
    /// <c>isDefault</c>). Migratsiya birorta shunday qator YOZMAYDI.
    /// </summary>
    public bool IsDefault { get; set; }

    /// <summary>false = yangi tanlovda ko'rinmaydi; qo'yilgan holatlar qoladi.</summary>
    public bool IsActive { get; set; } = true;

    public DateTimeOffset CreatedAt { get; set; }
}

File: schoollms.client/src/api/services/studentStatuses.ts
import { api, USE_MOCK } from '../client'

/* =========================================================================
 *  O'quvchi holati taglari — docs/modules/students-parity.md §2.3 (S-5).
 *
 *  Holat — ARXIVLASHDAN OLDINGI kuzatuv vositasi ("VIP", "Sinov muddatida",
 *  "Ko'chib ketmoqchi"). Arxivlash sabablari bilan chalkashtirmang: u
 *  o'quvchi KETGANDAN keyingi yorliq.
 * ========================================================================= */

/** Katalogdagi bitta holat. */
export interface StudentStatusTag {
  id: string
  name: string
  /** `#RRGGBB` yoki null (neytral rang). */
  color: string | null
  position: number
  /** Tizim qatori — tahrirlab ham, o'chirib ham bo'lmaydi. */
  isDefault: boolean
  isActive: boolean
  /** Nechta o'quvchida qo'yilgan — 0 bo'lsagina o'chirish mumkin. */
  usedBy: number
}

export interface SaveStudentStatusInput {
  name: string
  /** `#RRGGBB`; bo'sh satr — rangni tozalaydi. */
  color?: string | null
  position?: number
  isActive?: boolean
}

export async function getStudentStatuses(includeInactive = false): Promise<StudentStatusTag[]> {
  if (USE_MOCK) return []
  const { data } = await api.get<StudentStatusTag[]>('/admin/student-statuses', {
    params: includeInactive ? { includeInactive: true } : undefined,
  })
  return data
}

export async function createStudentStatus(input: SaveStudentStatusInput): Promise<StudentStatusTag> {
  const { data } = await api.post<StudentStatusTag>('/admin/student-statuses', input)
  return data
}

export async function updateStudentStatus(
  id: string,
  input: SaveStudentStatusInput,
): Promise<StudentStatusTag> {
  const { data } = await api.put<StudentStatusTag>(`/admin/student-statuses/${id}`, input)
  return data
}

export async function deleteStudentStatus(id: string): Promise<void> {
  await api.delete(`/admin/student-statuses/${id}`)
}

/** O'quvchiga holat qo'yadi. `null` — holatni olib tashlaydi. */
export async function setStudentStatus(studentId: string, statusId: string | null): Promise<void> {
  await api.put(`/admin/students/${studentId}/status`, { statusId })
}
